perf/src/client.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    net::{IpAddr, Ipv6Addr, SocketAddr},
+    sync::Arc,
+};
+
+use anyhow::{Context, Result};
+use bytes::Bytes;
+use futures::StreamExt;
+use structopt::StructOpt;
+use tokio::sync::Semaphore;
+use tracing::{error, info, trace};
+
+/// Connects to a QUIC perf server and maintains a specified pattern of requests until interrupted
+#[derive(StructOpt)]
+#[structopt(name = "client")]
+struct Opt {
+    /// Address to connect to
+    #[structopt(default_value = "localhost:4433")]
+    host: String,
+    /// Number of unidirectional requests to maintain concurrently
+    #[structopt(long, default_value = "0")]
+    uni_requests: u64,
+    /// Number of bidirectional requests to maintain concurrently
+    #[structopt(long, default_value = "1")]
+    bi_requests: u64,
+    /// Number of bytes to request
+    #[structopt(long, default_value = "1048576")]
+    request_size: u64,
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() {
+    let opt = Opt::from_args();
+
+    tracing_subscriber::fmt::init();
+
+    if let Err(e) = run(opt).await {
+        error!("{:#}", e);
+    }
+}
+
+async fn run(opt: Opt) -> Result<()> {
+    let addr = tokio::net::lookup_host(&opt.host)
+        .await
+        .context("resolving host")?
+        .next()
+        .unwrap();
+    let host_name = opt.host.split(':').next().unwrap();
+
+    info!("connecting to {} at {}", host_name, addr);
+
+    let endpoint = quinn::EndpointBuilder::default();
+
+    let (endpoint, _) = endpoint
+        .bind(&SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0))
+        .context("binding endpoint")?;
+
+    let mut cfg = quinn::ClientConfigBuilder::default();
+    cfg.protocols(&[b"perf"]);
+    let mut cfg = cfg.build();
+    let tls_cfg: &mut rustls::ClientConfig = Arc::get_mut(&mut cfg.crypto).unwrap();
+    tls_cfg
+        .dangerous()
+        .set_certificate_verifier(SkipServerVerification::new());
+
+    let quinn::NewConnection {
+        connection,
+        mut uni_streams,
+        ..
+    } = endpoint
+        .connect_with(cfg, &addr, &host_name)?
+        .await
+        .context("connecting")?;
+
+    info!("established");
+
+    tokio::spawn(async move {
+        while let Some(Ok(stream)) = uni_streams.next().await {
+            tokio::spawn(async move {
+                if let Err(e) = drain_stream(stream).await {
+                    error!("reading response failed: {:#}", e);
+                }
+            });
+        }
+    });
+
+    tokio::select! {
+        x = drive_uni(connection.clone(), opt.uni_requests, opt.request_size) => x?,
+        x = drive_bi(connection.clone(), opt.bi_requests, opt.request_size) => x?,
+        _ = tokio::signal::ctrl_c() => {
+            info!("shutting down");
+            connection.close(0u32.into(), b"interrupted");
+        }
+    }
+
+    endpoint.wait_idle().await;
+
+    // TODO: Print stats
```

Comment:
Yes, we will need those to make it useful. Probably similar to what exists in `bulk`